src/collapse/sample.rs

```diff
@@ -0,0 +1,226 @@
+        let func = line.next()?.trim_end();
+
+        let mut module = "";
+        if !self.opt.no_modules {
+            // Modules are shown with "(in libfoo.dylib)" or "(in AppKit)".
+            // We've arleady split on "(in " above.
+            if let Some(line) = line.next() {
+                if let Some(close) = line.find(')') {
+                    module = &line[..close];
+                }
+
+                // Remove ".dylib", since it adds no value.
+                if module.ends_with(".dylib") {
+                    module = &module[..module.len() - 6]
+                }
+            }
+        }
+
+        Some((time, func, module))
+    }
+
+    fn is_indent_char(c: char) -> bool {
+        c == ' ' || c == '+' || c == '|' || c == ':' || c == '!'
+    }
+
+    // Handle call graph lines of the form:
+    //
+    // 5130 Thread_8749954
+    //    + 5130 start_wqthread  (in libsystem_pthread.dylib) ...
+    //    +   4282 _pthread_wqthread  (in libsystem_pthread.dylib) ...
+    //    +   ! 4282 __doworkq_kernreturn  (in libsystem_kernel.dylib) ...
+    //    +   848 _pthread_wqthread  (in libsystem_pthread.dylib) ...
+    //    +     848 __doworkq_kernreturn  (in libsystem_kernel.dylib) ...
+    fn on_line<W: Write>(&mut self, line: &str, writer: &mut W) -> io::Result<()> {
+        if let Some(indent_chars) = line[4..].find(|c| !Self::is_indent_char(c)) {
+            // Each indent is two characters
+            if indent_chars % 2 != 0 {
+                error!("Odd number of indentation characters for line:\n{}", line);
+            }
+
+            let prev_depth = self.stack.len();
+            let depth = indent_chars / 2 + 1;
+
+            if depth <= prev_depth {
+                self.write_stack(writer)?;
+                for _ in 0..=prev_depth - depth {
+                    self.stack.pop();
+                }
+            } else if depth > prev_depth + 1 {
+                error!("Skipped indentation level at line:\n{}", line);
+            }
+
+            if let Some((samples, func, module)) = self.line_parts(&line[4 + indent_chars..]) {
+                if let Ok(samples) = samples.parse::<usize>() {
+                    self.current_samples = samples;
+                    if module.is_empty() {
+                        self.stack.push(func.to_owned());
+                    } else {
+                        self.stack.push(format!("{}`{}", module, func));
+                    }
+                } else {
+                    error!("Invalid samples field: {}", samples);
+                }
+            } else {
+                error!("Unable to parse stack line:\n{}", line);
+            }
+        } else {
+            error!("Found stack line with only indent characters:\n{}", line);
+        }
+
+        Ok(())
+    }
+
+    fn write_stack<W: Write>(&self, writer: &mut W) -> io::Result<()> {
+        if let Some(func) = self.stack.last() {
+            for symbol in IGNORE_SYMBOLS {
```

Comment:
How sure are we that everyone will always want to ignore these?

Reply:
Not sure at all. I was just following upstream. Maybe we should add an option for this?